sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/index/SupportsIndex.java

```diff
@@ -0,0 +1,139 @@
+   * Deleted index can be restored by calling restoreIndex.
+   *
+   * @param indexName the name of the index to be deleted
+   * @return true if the index is deleted
+   * @throws NoSuchIndexException If the index does not exist (optional)
+   * @throws UnsupportedOperationException If delete index is not a supported operation
+   */
+  default boolean deleteIndex(String indexName)
+      throws NoSuchIndexException, UnsupportedOperationException {
+    throw new UnsupportedOperationException("Delete index is not supported.");
+  }
+
+  /**
+   * Checks whether an index exists.
+   *
+   * @param indexName the name of the index
+   * @return true if the index exists, false otherwise
+   */
+  boolean indexExists(String indexName);
+
+  /**
+   * Lists all the indexes in a table.
+   *
+   * @param table the table to be checked on for indexes
+   * @throws NoSuchTableException
+   */
+  Index[] listIndexes(Identifier table) throws NoSuchTableException;
+
+  /**
+   * Hard deletes the index with the given name.
+   * The Index can't be restored once dropped.
+   *
+   * @param indexName the name of the index to be dropped.
+   * @return true if the index is dropped
+   * @throws NoSuchIndexException If the index does not exist (optional)
+   * @throws UnsupportedOperationException If drop index is not a supported operation
+   */
+  boolean dropIndex(String indexName) throws NoSuchIndexException, UnsupportedOperationException;
+
+  /**
+   * Restores the index with the given name.
+   * Deleted index can be restored by calling restoreIndex, but dropped index can't be restored.
+   *
+   * @param indexName the name of the index to be restored
+   * @return true if the index is restored
+   * @throws NoSuchIndexException If the index does not exist (optional)
+   * @throws UnsupportedOperationException
+   */
+  default boolean restoreIndex(String indexName)
+      throws NoSuchIndexException, UnsupportedOperationException {
+    throw new UnsupportedOperationException("Restore index is not supported.");
+  }
+
+  /**
+   * Refreshes index using the latest data. This causes the index to be rebuilt.
+   *
+   * @param indexName the name of the index to be rebuilt
+   * @return true if the index is rebuilt
+   * @throws NoSuchIndexException If the index does not exist (optional)
+   * @throws UnsupportedOperationException
+   */
+  default boolean refreshIndex(String indexName)
+      throws NoSuchIndexException, UnsupportedOperationException {
+    throw new UnsupportedOperationException("Refresh index is not supported.");
+  }
+
+  /**
+   * Alter Index using the new property. This causes the index to be rebuilt.
+   *
+   * @param indexName the name of the index to be altered
+   * @return true if the index is altered
+   * @throws NoSuchIndexException If the index does not exist (optional)
+   * @throws UnsupportedOperationException
+   */
+  default boolean alterIndex(String indexName, Properties properties)
```

Comment:
I think we should follow `TableCatalog.alterTable`
```
alterIndex(String indexName, IndexChange[] changes)
```

It's not very clear to me what can be changed in ALTER INDEX though.